include/fiction/algorithms/physical_design/apply_gate_library.hpp

```diff
@@ -146,9 +233,57 @@ CellLyt apply_gate_library(const GateLyt& lyt)
 
     detail::apply_gate_library_impl<CellLyt, GateLibrary, GateLyt> p{lyt};
 
-    auto result = p.run();
+    try
+    {
+        return p.run();
+    }
+    catch (const std::exception& e)
+    {
+        throw;
+    }
+}
+/**
+ * Applies a dynamic gate library (i.e., gate are designed on the fly by respecting atomic defects) to a given
+ * gate-level layout and, thereby, creates and returns a cell-level layout.
+ *
+ * May pass through, and thereby throw, an `unsupported_gate_type_exception` or an
+ * `unsupported_gate_orientation_exception`.
+ *
+ * @tparam CellLyt Type of the returned cell-level layout.
+ * @tparam GateLibrary Type of the gate library to apply.
+ * @tparam GateLyt Type of the gate-level layout to apply the library to.
+ * @tparam GateLibraryblack Type of the gate-level layout used to generate the blacklist.
+ * @param lyt The gate-level layout.
+ * @param defect_surface Defect surface with all atomic defects.
+ * @param params Parameter for the dynamic gate library.
+ * @param black_list Blacklist.
+ * @return A cell-level layout that implements `lyt`'s gate types with building blocks defined in `GateLibrary`.
+ */
+template <typename CellLyt, typename GateLibrary, typename GateLyt, typename GateLibraryblack>
+[[nodiscard]] CellLyt apply_dynamic_gate_library(
+    const GateLyt& lyt, const sidb_surface<CellLyt>& defect_surface, const sidb_dynamic_gate_library_params& params,
+    surface_black_list<
+        GateLyt, typename decltype(GateLibraryblack::get_gate_ports())::mapped_type::value_type::port_type>& black_list)
+{
+    static_assert(is_cell_level_layout_v<CellLyt>, "CellLyt is not a cell-level layout");
+    static_assert(!has_siqad_coord_v<CellLyt>, "CellLyt cannot have SiQAD coordinates");
+    static_assert(is_gate_level_layout_v<GateLyt>, "GateLyt is not a gate-level layout");
+    static_assert(mockturtle::has_is_constant_v<GateLyt>, "GateLyt does not implement the is_constant function");
+    static_assert(mockturtle::has_foreach_node_v<GateLyt>, "GateLyt does not implement the foreach_node function");
 
-    return result;
+    static_assert(std::is_same_v<technology<CellLyt>, technology<GateLibrary>>,
+                  "CellLyt and GateLibrary must implement the same technology");
+
+    detail::apply_gate_library_impl<CellLyt, GateLibrary, GateLyt> p{lyt};
+
+    try
+    {
+        return p.template run_dynamic_gates<GateLibraryblack>(defect_surface, params, black_list);
```

Comment:
```suggestion
        return p.run_dynamic_gates<GateLibraryblack>(defect_surface, params, black_list);
```

Reply:
@marcelwa Are you really sure? It does not work without.

Reply:
No, not certain. It was worth a shot. If it doesn't work without it, you can leave it of course.

Reply:
Absolutely! Many thanks!